### PR TITLE
GTJ-11

### PR DESCRIPTION
**Pull request by gulyaasjaanos**
db-migration changes: table name put into plurals
add building backend: buildings endpoint, repos and service add and resource repo changed
add building frontend: simple layout with buttons add

### CODE REVIEW

File: backend/migrations/20200801122150-buildings-add.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+var seed;
+
+exports.setup = function(options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+exports.up = function(db) {
+  return db.createTable("buildings", {
```
Comment (adamgyulavari) [inline backend/migrations/20200801122150-buildings-add.js:14]: @gulyaasjaanos pls use renameTable instead
https://db-migrate.readthedocs.io/en/latest/API/SQL/#renametabletablename-newtablename-callback

File: backend/src/controllers/buildingsController.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { buildingService } from '../services';
+
+const {errorMessages} = buildingService;
+
+const errorTypes = {
+    [errorMessages.missingType]: {status: 400, message: 'Type is required.'},
+    [errorMessages.wrongType]: {status: 404, message: 'Wrong type.'},
+    [errorMessages.notEnoughGold]: {status: 400, message: "You don't have enough money."},
+    [errorMessages.invalidId]: {status: 404, message: 'Invalid kingdom id.'}
```
Comment (adamgyulavari) [inline backend/src/controllers/buildingsController.js:9]: @gulyaasjaanos nice 👍

File: frontend/src/components/buildings/AddBuilding.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    const onErrorClick = () => {
+        if (error) setError(false);
+    };
+
+    return (
+        <section className='add_building'>
+            {buttons.map( (buildingData,i) => (
+                <AddBuildingButton key={i} buildingData={buildingData} goldAmount={goldAmount} onClick={onButtonClick}/>
+            ))}
+            <p className={error ? 'error visible' : 'error nonvisible'} onClick={onErrorClick} >Not enough gold.</p>
```
Comment (adamgyulavari) [inline frontend/src/components/buildings/AddBuilding.js:34]: @gulyaasjaanos just simply
```suggestion
            { error && <p className='error visible' onClick={onErrorClick} >Not enough gold.</p> }
```

File: backend/src/services/resourceService.js [hunk 4]
```diff
@@ -105,5 +67,43 @@ export const resourceService = {
...
+  ],
+  'food': [
+    [ kingdomId, cost, 0, resources.filter(e => e.type === 'gold')[0] ],
+    [ kingdomId, 0, genAdd, resources.filter(e => e.type === 'food')[0] ]
+  ],
+  'default': [
+    [ kingdomId, null, 0, resources.filter(e => e.type === 'gold')[0] ],
+    [ kingdomId, null, 0, resources.filter(e => e.type === 'food')[0]  ]
+  ]
+});
```
Comment (adamgyulavari) [inline backend/src/services/resourceService.js:94]: @gulyaasjaanos this looks like some weird magic, what's the point?
also `resources.filter(...)[0]` is `resources.find(...)`

File: backend/src/services/buildingService.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  invalidId: 4
+};
+
+const defaultData = {
+  farm : { hp : 1, cost : 100, time : 60000, gen : 5, genType : 'food' },
+  mine : { hp : 1, cost : 100, time : 60000, gen : 5, genType : 'gold' },
+  academy : { hp : 1, cost : 100, time : 60000, gen : 0, genType : '-' }
+}
+
+const buildingData = (data) => {
```
Comment (adamgyulavari) [inline backend/src/services/buildingService.js:18]: @gulyaasjaanos function names should be verbs usually, so `createBuildingData` maybe?

File: backend/migrations/20200801122150-buildings-add.js (hunk 1 above)
Comment (gulyaasjaanos) [reply backend/migrations/20200801122150-buildings-add.js:14]: 'buildings' is a new table

Comment (adamgyulavari) [reply backend/migrations/20200801122150-buildings-add.js:14]: oh, okay :D

File: frontend/src/components/buildings/AddBuilding.js (hunk 3 above)
Comment (gulyaasjaanos) [reply frontend/src/components/buildings/AddBuilding.js:34]: cool, thank you

File: backend/src/services/resourceService.js (hunk 4 above)
Comment (adamgyulavari) [reply backend/src/services/resourceService.js:94]: @gulyaasjaanos 
- generateResources
- spendGold
- updateGoldGeneration
- spendFood
- updateFoodGeneration

// buildingService - add Mine
resourcesService.spendGold(10)
resourcesService.updateFoodGeneration(1)

**Comment by adamgyulavari** [comment]
@gulyaasjaanos don't forget to fix the conflicts as well

File: backend/src/services/buildingService.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import { resourceService } from './resourceService';
+
+const errorMessages = {
+  missingType: 1,
+  wrongType: 2,
+  notEnoughGold: 3,
+  invalidId: 4
+};
+
+const defaultData = {
```
Comment (adamgyulavari) [inline backend/src/services/buildingService.js:12]: @gulyaasjaanos let's call this `buildingStats` instead

File: backend/src/services/buildingService.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+const validateKingdomId = async (id) => {
+  const kingdomBase = await kingdomRepo.getKingdomBaseData({'kingdom_id' : id});
+  if ( kingdomBase.length === 0 ) throw new Error(errorMessages.invalidId);
+};
+
+const validateBuild = async (id,type) => {
+  const resources = await resourceService.getResource(id);
+  const gold = resources.resources.filter(e => e.type === 'gold')[0].amount;
+  if ( gold < defaultData[type].cost ) throw new Error(errorMessages.notEnoughGold);
```
Comment (adamgyulavari) [inline backend/src/services/buildingService.js:56]: @gulyaasjaanos only this part is validating the building

File: backend/src/services/buildingService.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  return {
+    cost: defaultData[type].cost,
+    genAdd: defaultData[type].gen,
+    genType: defaultData[type].genType
+  };
```
Comment (adamgyulavari) [inline backend/src/services/buildingService.js:61]: @gulyaasjaanos this part is something like `getResourceChanges`

**Review by adamgyulavari — CHANGES_REQUESTED** [review]
other than that looks good 👍

**Comment by gulyaasjaanos** [comment]
I don't understand the fail in the check:
app-session.test.js is not involved in any recent changes
local tests are good
postman throws the proper error

**Review by adamgyulavari — APPROVED** [review]

**Pull request merged**